### ISSUE TITLE
Achieve covariance with immutability

### ISSUE THREAD

**Issue by cn7r66** [vimeo/psalm#2361]
Hi!

I don't understand how can I achieve covariance with immutability.

I tried the following example with a container class marked as @psalm-immutable where every mutable method return a new instance of the container instead of change the internal state.
I thought that was possible for the class to be covariant since if we pass a Collection<Dog> to Collection<Animal> and then add a new Cat() we correctly have a new Collection<Animal> with the original collection of Dogs in the same original state. However Psalm is complaining that we can't use a covariant template argument as input.

Complete example: [https://psalm.dev/r/21e7555056](https://psalm.dev/r/21e7555056)
```
/**
  * @psalm-immutable
  * @template-covariant T
  */
class Collection {

    /** @var list<T>*/
    private $arr = [];
  
   /** 
      * @param T ...$a
      */
    public function __construct(...$a) {
    	$this->arr = $a;
    }
  
   /**
     * @param T $a
     * @return Collection<T>
     */
    public function add($a) : Collection
    {
      return new Collection(...$this->arr, $a);
    }
}
```

My next try was to extract the method and put the same body inside a function but Psalm can't correctly infer the type of T and let me assign a new Cat() to a Collection<Dog> without complaining, don't know if I'm using the templates bad or this is a bug.

Complete example: [https://psalm.dev/r/a0caa3593d](https://psalm.dev/r/a0caa3593d)
```
/**
  * @template T
  * @param Collection<T> $c
  * @param T $a
  * @return Collection<T>
  */
function add(Collection $c, $a) : Collection
{
  return new Collection(...$c->toArray(), $a);
}

/**
  * @param Collection<Dog> $c
  */
function acceptDogs(Collection $c) : void{}

$dogs = add($dogs, new Cat()); //Shouldn't this throw an error?
acceptDogs($dogs);
```
Since this solution is too implementation dependent I tried with two separate interfaces, the first read-only and the second with write methods extending the first.

Complete example: [https://psalm.dev/r/2d09fdbeeb](https://psalm.dev/r/2d09fdbeeb)
```
/**
  * @template-covariant T
  */
interface Collection
{
  /**
    * @return list<T>
    */
   public function toArray() : array;
}

/**
  * @template T
  * @template-extends Collection<T>
  */
interface WritableCollection extends Collection
{
    /**
      * @param T $a
      * @return WritableCollection<T>
      */
     public function add($a) : WritableCollection;
}
```

With this everything is ok, works as expected and Psalm can correctly infer types with covariance/invariance.

Although this solution seems acceptable to me i think that an immutable class can be allowed to take a template argument as input and still be covariant.

Is this possible or I'm missing something?

**Comment by muglug** [comment]
> Although this solution seems acceptable to me i think that an immutable class can be allowed to take a template argument as input and still be covariant.

Yeah, though it's an edge-case that seems reasonable - @Ocramius do you agree?

**Comment by Ocramius** [comment]
I don't fully understand `@template-covariant` TBH (and why it is needed: will need to read up on it).

As for the example above, state being retained must be immutable to be part of an immutable interface: are `Cat` and `Dog` immutable? Or did I find a completely unrelated bug?

**Comment by cn7r66** [comment]
> As for the example above, state being retained must be immutable to be part of an immutable interface: are `Cat` and `Dog` immutable? Or did I find a completely unrelated bug?

This is a very good point!

I was looking into the @psalm-immutable tag family [here](https://psalm.dev/articles/immutability-and-beyond) but a behavior like that does not seem to be supported.

Maybe adding some additional information in the method docblock to point out that an argument must be marked as @psalm-immutable? 

Something like this:

```
   /**
     * @param T $a
     *
     * @psalm-immutable-param $a
     *
     * @return Collection<T>
     */
    public function add($a) : Collection
    {
      return new Collection(...$this->arr, $a);
    }
```

**Comment by jesseschalken** [comment]
>Although this solution seems acceptable to me i think that an immutable class can be allowed to take a template argument as input and still be covariant.

Being immutable isn't sufficient. Example:

```php
abstract class Animal {}
class Dog extends Animal {
  function bark(): void {}
}
class Cat extends Animal {
  function meow(): void {}
}

/** @template T */
interface Task {
  /** @param T $x */
  function run($x): void;
}

/** @implements Task<Dog> */
class DogTask implements Task {
  function run($x): void {
    $x->bark();
  }
}

/** @param Task<Animal> $t */
function runAnimalTask($t): void {
  $t->run(new Cat());
}

function main(): void {
  runAnimalTask(new DogTask()); // will try to call ->bark() on a Cat
}
```

Your top level function `add` looks fine to me. I think the lack of error on `$dogs = add($dogs, new Cat()); acceptDogs($dogs);` is a bug.

To have a type safe `add` *method* you would need a `super` type constraint. See https://hhvm.com/blog/9215/covariance-contravariance-and-super-type-constraints

**Comment by cn7r66** [comment]
> To have a type safe `add` _method_ you would need a `super` type constraint. See https://hhvm.com/blog/9215/covariance-contravariance-and-super-type-constraints

Oh, I see! That make sense now, thanks!

So, without the `super` keyword, what about an invariant interface extending a covariant one like in my previous example?

Intuitively seems ok to me (while the opposite is not) and even Psalm has nothing to complain, but if I do this, it raises an error:

[https://psalm.dev/r/aab218dd1d](https://psalm.dev/r/aab218dd1d)
```
/**
  * @template T
  * @template-extends IteratorAggregate<int, T>
  */
interface Invariant extends IteratorAggregate {
  
    /**
      * @param T $t
      */
	public function write($t):void{}
}
```
```
ERROR: InvalidTemplateParam - 10:16 - Template param T of Invariant is marked covariant and cannot be used as input to a function
```
However if I extends IteratorAggregate first to a covariant interface and then to an invariant one Psalm raises no error:

[https://psalm.dev/r/cc24e6a482](https://psalm.dev/r/cc24e6a482)
```
/**
  * @template-covariant T
  * @template-extends IteratorAggregate<int, T>
  */
interface Covariant extends IteratorAggregate
{
  /**
    * @return T
    */
  public function read(){}
}

/**
  * @template T
  * @template-extends Covariant<T>
  */
interface Invariant extends Covariant
{
    /**
      * @param T $a
      */
      public function write($a) : void {}
} 
```
Maybe it has to do with the fact that the first template argument of `IteratorAggregate` was setted to int?


On the opposite, if a covariant interface extends an invariant one i get no error on interface level

[https://psalm.dev/r/e0a5194097](https://psalm.dev/r/e0a5194097)
```
/** @template T */
interface Writable {
  /** @param T $x */
  function write($x): void;
}

/**
  * @template-covariant T
  * @template-extends Writable<T>
  */
interface Readable extends Writable {
  /** @return T */
  public function read() {}
}
```

But if I add an implementation it raises an error on `Readable` T param

[https://psalm.dev/r/e95702f0a6](https://psalm.dev/r/e95702f0a6)
```
/** @implements Readable<Dog> */
class DogTask implements Readable {
  function write($x): void {
    $x->bark();
  }
  
  function read() {
    return new Dog();
  }
}
```

```
ERROR: InvalidTemplateParam - 13:14 - Template param T of Readable is marked covariant and cannot be used as input to a function

INFO: MixedMethodCall - 29:9 - Cannot determine the type of $x when calling method bark
```

Why the error can't be raised without an implementation? There are some cases where an interface hierarchy like that can have a type safe implementation?

**Comment by jesseschalken** [comment]
> So, without the super keyword, what about an invariant interface extending a covariant one like in my previous example?
>
> Intuitively seems ok to me (while the opposite is not) and even Psalm has nothing to complain, but if I do this, it raises an error.

I agree, this sounds like a bug. It works in Kotlin, FWIW: https://pl.kotl.in/kECkQhunL.

**Comment by cn7r66** [comment]
> I agree, this sounds like a bug

I think so too.

I made some tests and it seems that the problem arise only when an invariant class partially specializes a covariant one with multiple template arguments.

Summary:

ok `Covariant<K, V> ← Covariant<K, V>`  Example: [https://psalm.dev/r/0ad3b4478e](https://psalm.dev/r/0ad3b4478e)

ok `Invariant<K, V> ← Covariant<K, V>` Example: [https://psalm.dev/r/bc4de25d41](https://psalm.dev/r/bc4de25d41)

ok `Covariant<specialized, specialized> ← Covariant<K, V>` Example: [https://psalm.dev/r/4f53702a02](https://psalm.dev/r/4f53702a02)

ok `Invariant<specialized, specialized> ← Covariant<K, V>` Example: [https://psalm.dev/r/5532af06be](https://psalm.dev/r/5532af06be)

ok `Covariant<specialized, V> ← Covariant<K, V>` Example: [https://psalm.dev/r/0b89da2404](https://psalm.dev/r/0b89da2404)


Error `Invariant<specialized, V> ← Covariant<K, V>` Example: [https://psalm.dev/r/0db6815cdb](https://psalm.dev/r/0db6815cdb)

```
ERROR: InvalidTemplateParam - 27:15 - Template param V of Collection is marked covariant and cannot be used as input to a function
```

Shouldn't `V` still be covariant on interface level and invariant on class level?
I would not expect an error on line 27 but two errors on lines 78 and 79 (because of the invariance).

**Comment by jesseschalken** [comment]
> Shouldn't `V` still be covariant on interface level and invariant on class level?

Why? Classes and interfaces are both types that contain method signatures, and covariant/contravariance is entirely about how subtyping of types containing method signatures relates to the subtyping of the inputs and outputs of those methods.

That said you could have an error on line 78 and 79 instead of line 27 using [use-site variance annotations](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)#Use-site_variance_annotations_(wildcards)) like Java instead of declaration-site variance annotations.

**Comment by muglug** [comment]
The original issue has been fixed in 532e2d6. I suggest ticketing the additional issue(s) separately. I don't totally have my head around all the covariant/contravariant rules, but if you set out how Psalm _should_ work I should be able to implement it.

If a template type is marked covariant, currently that covariance is propagated at least one level down if there are `@template-extends Foo<T>` annotations. Maybe that's not correct – but it felt good at the time.
